homeassistant/components/workday/binary_sensor.py

```diff
@@ -76,43 +81,64 @@ def valid_country(value: Any) -> str:
 )
 
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
-    add_entities: AddEntitiesCallback,
+    async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the Workday sensor."""
-    add_holidays: list[DateLike] = config[CONF_ADD_HOLIDAYS]
-    remove_holidays: list[str] = config[CONF_REMOVE_HOLIDAYS]
-    country: str = config[CONF_COUNTRY]
-    days_offset: int = config[CONF_OFFSET]
-    excludes: list[str] = config[CONF_EXCLUDES]
-    province: str | None = config.get(CONF_PROVINCE)
-    sensor_name: str = config[CONF_NAME]
-    workdays: list[str] = config[CONF_WORKDAYS]
-
-    year: int = (get_date(dt.now()) + timedelta(days=days_offset)).year
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.6.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={
+                **config,
+                CONF_PROVINCE: config.get(CONF_PROVINCE),
+            },
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Workday sensor."""
+    add_holidays: list[DateLike] = entry.options[CONF_ADD_HOLIDAYS]
+    remove_holidays: list[str] = entry.options[CONF_REMOVE_HOLIDAYS]
+    country: str = entry.options[CONF_COUNTRY]
+    days_offset: int = int(entry.options[CONF_OFFSET])
+    excludes: list[str] = entry.options[CONF_EXCLUDES]
+    province: str | None = entry.options.get(CONF_PROVINCE)
+    sensor_name: str = entry.options[CONF_NAME]
+    workdays: list[str] = entry.options[CONF_WORKDAYS]
+
+    year: int = (dt.now() + timedelta(days=days_offset)).year
     obj_holidays: HolidayBase = getattr(holidays, country)(years=year)
 
-    if province:
-        if (
-            hasattr(obj_holidays, "subdivisions")
-            and province in obj_holidays.subdivisions
-        ):
+    if province and province != "None":
+        try:
             obj_holidays = getattr(holidays, country)(subdiv=province, years=year)
-        else:
+        except NotImplementedError:
             LOGGER.error("There is no subdivision %s in country %s", province, country)
             return
 
     # Add custom holidays
-    try:
+    if add_holidays != []:
         obj_holidays.append(add_holidays)
-    except TypeError:
-        LOGGER.debug("No custom holidays or invalid holidays")
```

Comment:
I think the try...except can be removed as a preliminary PR.
The schema validates it as a list of string, so it's probably a leftover from a previous refactor?

Reply:
This can't be removed until we have a validation in place which comes with config_flow so needs to stay in this PR

Reply:
Was that not part of the YAML validation?
Maybe it was the tests that were wrong?

```python
        vol.Optional(CONF_ADD_HOLIDAYS, default=[]): vol.All(
            cv.ensure_list, [cv.string]
        ),
        vol.Optional(CONF_REMOVE_HOLIDAYS, default=[]): vol.All(
            cv.ensure_list, [cv.string]
        ),
```

Reply:
That validation is of course there. I meant more if a user puts `2023-12-32` or some date which doesn't exist.
That validation will be implemented as per config flow and therefore it needs to stay for the moment.